### PR TITLE
Extensively use Numpy for similarity based recommendations

### PR DESCRIPTION
**Pull request by Dexterp37**
This PR will extract Numpy arrays/matrices out of the donor
and LR curves files. All the heavy lifting computations
were changed to make use of the vectorized Numpy facilities.

### CODE REVIEW

**Comment by Dexterp37** [comment]
This sacrificed "debuggability" (yes, I make up my own words!) to drop the time it takes to recommend fom 500ms to 40ms.

File: taar/recommenders/similarity_recommender.py [hunk 1]
```diff
@@ -28,30 +29,46 @@ class SimilarityRecommender(BaseRecommender):
...
+        """This function build two feature cache matrices.
+
+        One matrix is for the continuous features and the other is for
+        the categorical features. This is needed to speed up the similarity
+        recommendation process."""
+
+        # Build a numpy matrix cache for the continuous features.
+        self.continuous_features = np.zeros((self.num_donors, len(CONTINUOUS_FEATURES)))
+        for idx, d in enumerate(self.donors_pool):
+            featuers = [d.get(specified_key) for specified_key in CONTINUOUS_FEATURES]
```
Comment (mlopatka) [inline taar/recommenders/similarity_recommender.py]: nit: typo in `features`

File: taar/recommenders/similarity_recommender.py [hunk 2]
```diff
@@ -28,30 +29,46 @@ class SimilarityRecommender(BaseRecommender):
...
+            self.continuous_features[idx] = featuers
+
+        # Build the cache for categorical features.
+        self.categorical_features =\
+            np.zeros((self.num_donors, len(CATEGORICAL_FEATURES)), dtype='object')
+        for idx, d in enumerate(self.donors_pool):
+            featuers = [d.get(specified_key) for specified_key in CATEGORICAL_FEATURES]
```
Comment (mlopatka) [inline taar/recommenders/similarity_recommender.py]: nit: typo in `features`

**Comment by Dexterp37** [comment]
> When you say "cached features" do you mean that those features can be cached and be reused for different users requesting recommendations?

Yes, we build up a Numpy matrix out of the data files that gets shared across all the recommendation requests

**Comment by maurodoglio** [comment]
Great, where is it cached? on disk?

**Comment by Dexterp37** [comment]
> Great, where is it cached? on disk?

Nope, that's stored in memory - `self.categorical_features` and `self.continuous_features`. There's no much use in storing it to the disk, as it's very fast to build it out of the data files (it's basically converting a subset of the fields into Numpy structures).

**Pull request merged**